Source/Fuse.Reactive.JavaScript/TreeObservable/TreeArray.uno

```diff
@@ -0,0 +1,193 @@
+using Uno.Collections;
+using Uno;
+using Fuse;
+
+namespace Fuse.Reactive
+{
+	class TreeArray : ArrayMirror, IObservableArray
+	{
+		internal TreeArray(Scripting.Array arr): base(arr) {}
+
+		public IDisposable Subscribe(IObserver observer)
+		{
+			return new ArraySubscription(this, observer);
+		}
+
+		internal class ArraySubscription: Subscription, ISubscription
+		{
+			readonly IObserver _observer;
+
+			public ArraySubscription(ArrayMirror am, IObserver observer): base(am)
+			{
+				_observer = observer;
+			}
+
+			public void OnReplaceAt(int index, object newValue)
+			{
+				_observer.OnNewAt(index, newValue);
+				var next = Next as ArraySubscription;
+				if (next != null) next.OnReplaceAt(index, newValue);
+			}
+
+			public void OnAdd(object newValue)
+			{
+				_observer.OnAdd(newValue);
+				var next = Next as ArraySubscription;
+				if (next != null) next.OnAdd(newValue);
+			}
+
+			public void OnInsertAt(int index, object newValue)
+			{
+				_observer.OnInsertAt(index, newValue);
+				var next = Next as ArraySubscription;
+				if (next != null) next.OnInsertAt(index, newValue);
+			}
+
+			public void OnRemoveAt(int index)
+			{
+				_observer.OnRemoveAt(index);
+				var next = Next as ArraySubscription;
+				if (next != null) next.OnRemoveAt(index);
+			}
+
+			public void OnReplaceAll(IArray values, ArraySubscription exclude)
+			{
+				if(this != exclude) _observer.OnNewAll(values);
+
+				var next = Next as ArraySubscription;
+				if(next != null) next.OnReplaceAll(values, exclude);
+			}
+
+			class ReplaceAllOperation
+			{
+				ThreadWorker _worker;
+				Scripting.Array _target;
+				IArray _newValues;
+
+				public ReplaceAllOperation(ThreadWorker worker, Scripting.Array target, IArray newValues)
+				{
+					_worker = worker;
+					_target = target;
+					_newValues = newValues;
+				}
+
+				public void Perform()
+				{
+					var ctx = _worker.Context;
+
+					var nv = new object[_newValues.Length];
+					for(var i = 0; i < _newValues.Length; ++i) {
+						nv[i] = _worker.Unwrap(_newValues[i]);
+					}
+					var newValuesJs = ctx.NewArray(nv);
+
+					var replaceAllFn = (Scripting.Function) ctx.Evaluate("replaceAll",
+						"(function(array, values) {" +
+							"if('__fuse_replaceAll' in array) array.__fuse_replaceAll(values);" +
+							"else {"+
+								"array.length = 0;"+
+								"Array.prototype.push.apply(array, values);"+
+							"}"+
+						"})");
```

Comment:
Looks like this one is constant, do we really need to re-create it every time?

Reply:
We definitely don't, although it needs to be evaluated in the right context. What's the correct place for this in your opinion? Builtins?

Reply:
TBH, I'm not really sure.

Reply:
Lets just do that as a follow-up to get things moving.